### PR TITLE
Feature/d3 asim 2894

### PR DESCRIPTION
**Pull request by spyrostz**

### CODE REVIEW

File: src/d3a/models/strategy/load_hours.py [hunk 1]
```diff
@@ -164,7 +164,7 @@ def _delete_past_state(self):
                 to_delete.append(k)
         for k in to_delete:
             del self.energy_requirement_Wh[k]
-            del self.state.desired_energy_Wh[k]
+            self.state.desired_energy_Wh.pop(k, None)
```
Comment (hannesdiedrich) [inline src/d3a/models/strategy/load_hours.py]: weird that this is needed. There should always be a `desired_energy_Wh` once the state was initiated. Right?

Comment (spyrostz) [reply src/d3a/models/strategy/load_hours.py]: Should not be needed. Some of the commits here are before the hackathon where we fixed these issues while debugging Canary network, that's why these should not be valid anymore.

Comment (hannesdiedrich) [reply src/d3a/models/strategy/load_hours.py]: AH, OK, I was browsing through the commits. Sorry.

File: src/d3a/models/state.py [hunk 2]
```diff
@@ -49,19 +50,81 @@ def get_state(self):
...
+            return self._energy_requirement_Wh[time_slot]
+        else:
+            return offer_energy_Wh
+
+    def calculate_energy_to_bid(self, time_slot):
+        return self._energy_requirement_Wh[time_slot]
+
+    def decrement_energy_requirement(self, purchased_energy_Wh, time_slot, area_name):
+        self._energy_requirement_Wh[time_slot] -= purchased_energy_Wh
+        assert self._energy_requirement_Wh[time_slot] >= -FLOATING_POINT_TOLERANCE, \
```
Comment (hannesdiedrich) [inline src/d3a/models/state.py:93]: Great that this assertion is now bound to the subtraction!

File: src/d3a/models/strategy/storage.py [hunk 3]
```diff
@@ -589,9 +589,10 @@ def event_offer(self, *, market_id, offer):
             market = self.area.get_future_market_from_id(market_id)
             # sometimes the offer event arrives earlier than the market_cycle event,
             # so the default values have to be written here too:
+            # TODO: Evaluate for removal
```
Comment (hannesdiedrich) [inline src/d3a/models/strategy/storage.py:592]: Indeed!

File: src/d3a/models/strategy/update_frequency.py [hunk 4]
```diff
@@ -48,6 +48,17 @@ def __init__(self, initial_rate, final_rate, fit_to_limit=True,
         self.number_of_available_updates = 0
         self.rate_limit_object = rate_limit_object
 
+    def delete_past_state_values(self, current_market_time_slot):
```
Comment (hannesdiedrich) [inline src/d3a/models/strategy/update_frequency.py:51]: Great!

**Review by hannesdiedrich — DISMISSED** [review]
Thanks for the refactoring! Looking forward to the PV and Storage refactoring

File: src/d3a/models/strategy/storage.py (hunk 3 above)
Comment (spyrostz) [reply src/d3a/models/strategy/storage.py:592]: I just recalled that I had added it there! I'll actually deal with it in this PR

**Review by faizan2590 — DISMISSED** [review]
LGTM

**Review by hannesdiedrich — APPROVED** [review]
LGTM

**Review by faizan2590 — APPROVED** [review]
LGTM

**Pull request merged**